drivers/wireless/lpwan/rn2xx3/rn2xx3.c

```diff
@@ -0,0 +1,1508 @@
+  int err;
+  char response[10];
+
+  err = file_write(&priv->uart, "radio get bw\r\n",
+                   sizeof("radio get bw\r\n") - 1);
+  if (err < 0)
+    {
+      return err;
+    }
+
+  err = read_line(priv, response, sizeof(response));
+
+  if (err < 0)
+    {
+      return err;
+    }
+
+  response[err] = '\0';
+  *bw = strtoul(response, NULL, 10);
+  return 0;
+}
+
+/****************************************************************************
+ * Name: rn2xx3_setsf
+ *
+ * Description:
+ *    Set the spread factor of the RN2xx3.
+ *
+ ****************************************************************************/
+
+static int rn2xx3_setsf(FAR struct rn2xx3_dev_s *priv, uint8_t sf)
+{
+  char sfstr[15];
+  ssize_t length;
+
+  length =
+      file_write(&priv->uart, "radio set sf ", sizeof("radio set sf ") - 1);
+  if (length < 0)
+    {
+      return length;
+    }
+
+  /* Write actual parameter and end command */
+
+  length = snprintf(sfstr, sizeof(sfstr), "sf%u\r\n", sf);
+  length = file_write(&priv->uart, sfstr, length);
+  if (length < 0)
+    {
+      return length;
+    }
+
+  return get_command_err(priv);
+}
+
+/****************************************************************************
+ * Name: rn2xx3_getsf
+ *
+ * Description:
+ *    Get the spread factor of the RN2xx3.
+ *
+ ****************************************************************************/
+
+static int rn2xx3_getsf(FAR struct rn2xx3_dev_s *priv, FAR uint8_t *sf)
+{
+  int err;
+  char response[20];
+
+  err = file_write(&priv->uart, "radio get sf\r\n",
+                   sizeof("radio get sf\r\n") - 1);
+  if (err < 0)
+    {
+      return err;
+    }
+
+  err = read_line(priv, response, sizeof(response));
+
+  if (err < 0)
+    {
+      return err;
+    }
+
+  response[err] = '\0';
+  *sf = strtoul(&response[2], NULL, 10); /* Skip 'sf' */
+  return 0;
+}
+
+/****************************************************************************
+ * Name: rn2xx3_setprlen
+ *
+ * Description:
+ *    Set the preamble length of the RN2xx3.
+ *
+ ****************************************************************************/
+
+static int rn2xx3_setprlen(FAR struct rn2xx3_dev_s *priv, uint16_t prlen)
+{
+  char prstr[16];
+  ssize_t length;
+
+  length = file_write(&priv->uart, "radio set prlen ",
+                      sizeof("radio set prlen ") - 1);
+  if (length < 0)
+    {
+      return length;
+    }
+
+  /* Write actual parameter and end command */
+
+  length = snprintf(prstr, sizeof(prstr), "%u\r\n", prlen);
+  length = file_write(&priv->uart, prstr, length);
+  if (length < 0)
+    {
+      return length;
+    }
+
+  return get_command_err(priv);
+}
+
+/****************************************************************************
+ * Name: rn2xx3_getprlen
+ *
+ * Description:
+ *    Get the preamble length of the RN2xx3.
+ *
+ ****************************************************************************/
+
+static int rn2xx3_getprlen(FAR struct rn2xx3_dev_s *priv,
```

Comment:
There is a lot of very similar code. Why not consolidate it all in one function, like

```
int rn2xx3_read_param(FAR struct rn2xx3_dev_s *priv, const char *req, char *resp, size_t resplen);
```

And just call it in those functions

```
if (rn2xx3_read_param(&priv->uart, "radio get prlen\r\n", resp, sizeof(resp) == -1)
   return err;

prlen = strtoul(resp, NULL, 10);
```

Reply:
I misread, I can factor out all the call and response before the parsing. Thanks!